src/backend/cpu/cast.hpp

```diff
@@ -26,6 +26,112 @@ struct UnOp<To, Ti, af_cast_t> {
     }
 };
 
+/// NOTE(umar): The next specializations have multiple eval functions because
+/// the f16 data type needs to be converted to and from the compute type.
+/// Here we have speciliztions for real numbers as well as the complex
+/// numbers
+/// TODO(umar): make a macro to reduce repeat code
+
+template<typename To>
+struct UnOp<To, common::half, af_cast_t> {
+    typedef common::half Ti;
+
+    void eval(jit::array<To> &out, const jit::array<Ti> &in, int lim) {
+        for (int i = 0; i < lim; i++) {
+            float val = in[i];
+            out[i]    = To(val);
+        }
+    }
+
+    void eval(jit::array<To> &out, const jit::array<float> &in, int lim) {
+        for (int i = 0; i < lim; i++) { out[i] = To(in[i]); }
+    }
+};
+
+template<typename Ti>
+struct UnOp<common::half, Ti, af_cast_t> {
+    typedef common::half To;
+
+    void eval(jit::array<To> &out, const jit::array<Ti> &in, int lim) {
+        for (int i = 0; i < lim; i++) {
+            float val = in[i];
+            out[i]    = To(val);
+        }
+    }
+
+    void eval(jit::array<float> &out, const jit::array<Ti> &in, int lim) {
+        for (int i = 0; i < lim; i++) { out[i] = float(in[i]); }
+    }
+};
+
+template<>
+struct UnOp<std::complex<float>, common::half, af_cast_t> {
+    typedef common::half Ti;
+    typedef std::complex<float> To;
+
+    void eval(jit::array<To> &out, const jit::array<Ti> &in, int lim) {
+        for (int i = 0; i < lim; i++) {
+            float val = in[i];
+            out[i]    = To(val);
+        }
+    }
+
+    void eval(jit::array<To> &out, const jit::array<float> &in, int lim) {
+        for (int i = 0; i < lim; i++) { out[i] = To(in[i]); }
+    }
+};
```

Comment:
The specialization for `std::complex<float>` output type for `struct UnOp<To, common::half, af_cast_t>` looks exactly same in code to the general implementation. Doesn't look like we need it from `half` to `To`.